src/lib.rs

```diff
@@ -1881,52 +1881,87 @@ extern "C" fn fn_derive_key(
         },
     }
 
-    let result = operation.derive(
+    let mut result = res_or_ret!(operation.derive(
         &key,
         tmpl,
         token.get_mechanisms(),
         token.get_object_factories(),
-    );
-    let (kh, addtl) = match result {
-        Ok((obj, addtl)) => {
-            let h = res_or_ret!(token.insert_object(s_handle, obj));
-            (h, addtl)
-        }
-        Err(e) => return err_to_rv!(e),
-    };
+    ));
+    if result.len() == 0 {
+        return CKR_GENERAL_ERROR;
+    }
 
-    let mut rv = CKR_OK;
-    if addtl > 0 {
-        let mut o = Vec::<CK_OBJECT_HANDLE>::with_capacity(addtl);
-        for _ in 0..addtl {
-            let r = operation.derive_additional_key();
-            match r {
-                Ok((obj, hptr)) => match token.insert_object(s_handle, obj) {
-                    Ok(h) => {
-                        unsafe { core::ptr::write(hptr as *mut _, h) };
-                        o.push(h);
+    match mechanism.mechanism {
+        CKM_SP800_108_COUNTER_KDF
+        | CKM_SP800_108_FEEDBACK_KDF
+        | CKM_SP800_108_DOUBLE_PIPELINE_KDF => {
```

Comment:
Does it make sense to keep this `CKM_SP800_108_DOUBLE_PIPELINE_KDF` in all the matches when they fail with the same way as the `_` case. Or are we close to the plan to implement this from scratch?

Reply:
I want to implement it from scratch at some point, so this will remind me of all places where I have to go and fix stuff later.